### ISSUE TITLE
Slow ListView row rendering

### ISSUE THREAD

**Issue by mcz** [facebook/react-native#5034]
Hello Everyone!

Here is an interesting problem regarding slow rendering. I got exactly the same problem as Drew Goodwin.

http://stackoverflow.com/questions/34073819/why-the-slow-screen-redraw-under-certain-circumstances-in-react-native

Even if I have only 30 rows in a ListView and I create a datasource

```
dataSource = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
```

Then I change the datasource:

```
dataSource = this.dataSource.cloneWithRows(excercises);
```

Rerendering the ListView is extremely slow. Just as you could see on the animated GIF on stackoverflow.

Any ideas are more than welcome.

Thank you!

**Comment by mcz** [comment]
In case it helps: Here is a Systrace HTML file when my slow ListView is being rendered.

[dump_1451429776399.html.zip](https://github.com/facebook/react-native/files/74454/dump_1451429776399.html.zip)

**Comment by hosainnet** [comment]
Have you tried adding 'initialListSize' on the ListView? I found adding it improved performance quite a lot, If you know the minimum list size beforehand.

**Comment by mcz** [comment]
@hosainnet Thanks for the idea. I tried it, but it didn't help at all. For me (according to the systrace) it looks like RN is rendering each row separately instead of rendering all rows in a batch ... even if I supply the initialListSize parameter.

**Comment by hosainnet** [comment]
@mcz from the docs:

"Rate-limited row rendering - By default, only one row is rendered per event-loop (customizable with the pageSize prop). This breaks up the work into smaller chunks to reduce the chance of dropping frames while rendering rows."

Try setting pageSize to something like 10 as well

**Comment by brentvatne** [comment]
pageSize is indeed useful for this too. If you want to force it to re-render with the initial list size, you can give the ListView a key and then change that key when you swap out with a brand new datasource.

**Comment by mcz** [comment]
@hosainnet and @brentvatne thank you for the ideas. I will check these out and let you know. For the time being I removed the whole ListView and started using ScrollView with multiple items inside of. It renders extremely fast within a blink of eye :) Interestingly much faster than with the ListView

**Comment by sunglee** [comment]
@mcz, did ScrollView make better rendering result? I had the same issue with ListView. When I selected an item from listview, JS thread value in performance monitor shows negative value (it shows -2.x fps) before transitioning to next screen. However UI thread stayed almost same as 60fps. So I dropped ListView and replaced it with ScrollView and removed all elements for the row to show only single TextView, but JS Thread FPS didn't improve at all. I am not quite sure where JS thread shows poor performance.
